### ISSUE TITLE
`AssetServer` and `gltF` workflows are very prone to failing silently

### ISSUE THREAD

**Issue by alice-i-cecile** [bevyengine/bevy#7581]
## What problem does this solve or what need does it fill?

When working with `gltF` files, there are quite a few ways for something to go wrong:

1. The file doesn't exist.
2. The path is wrong.
3. You forgot to specify `#Scene0`.
4. You specified the wrong scene number.
5. Your scale is wildly wrong.
6. Your model is inside out and you have backface culling enabled.
7. Your camera isn't looking at the model.
8. Your material is fully transparent.
9. You're attempting to save the `gltF` scene as something else, possibly a `Mesh` or `StandardMaterial`, rather than a `HandleScene`.
10. You specified an invalid scene path.

Currently, we detect and report 1 and 2. All other cases silently fail, with nothing (apparently) being rendered.

## What solution would you like?

Add error reporting and failures to cases that are flatly wrong: 3, 8, 9, 10.
Warn in cases that are possibly wrong: 4.
Add a prominently documented model debugging tool to look for and report on the remaining cases related to a malformed model.

## What alternative(s) have you considered?

The categorization of these failure modes is very much up for debate.

**Comment by rparrett** [comment]
I was surprised about 3/4, because I remembered some sort of error handling.

However, it seems like we intentionally ignore an error, presumably to support async loading?

https://github.com/bevyengine/bevy/blob/fefe5297ad00f93cc92df61153cb15dad3ae4857/crates/bevy_scene/src/scene_spawner.rs#L252-L257

**Comment by mockersf** [comment]
°1 and °2 should already log an error
°3 is because there is no way to know in advance the kind of asset that will be returned. "filename.gltf" could very well return an handle to a scene. This is because assets are loaded asynchronously, and we don't know what the file contains until after it's loaded
°4, °5, °6, °7 and °8 are using a "standard" definition of wrong which may not be true in all cases. Scene viewer tries to be smart for most of those in how to place the camera, but people using gltf files in their game should know their size, their scale, where they place them, ... a material fully transparent may be exactly what I want and I'll lerp my material to not transparent in a pretty way during runtime. spawning something off camera that will move after is a normal use case
°9 is equivalent to °3
I'm not sure what you mean by °10? By invalid scene path, you mean something invalid after the `#`? That can't be known until after the asset is loaded, but we can log something in that case.
For °11, if a system isn't running, Rust should already give a warning for something like "unused function"?

**Comment by alice-i-cecile** [comment]
> °3 is because there is no way to know in advance the kind of asset that will be returned. "filename.gltf" could very well return an handle to a scene. This is because assets are loaded asynchronously, and we don't know what the file contains until after it's loaded
> °9 is equivalent to °3

Disagreed here: we can check if the generated handles are to the correct type of asset upon loading. This also seems like something that could be checked based on the extension.

> °4, °5, °6, °7 and °8 are using a "standard" definition of wrong which may not be true in all cases. Scene viewer tries to be smart for most of those in how to place the camera, but people using gltf files in their game should know their size, their scale, where they place them, ... a material fully transparent may be exactly what I want and I'll lerp my material to not transparent in a pretty way during runtime. spawning something off camera that will move after is a normal use case

Agreed: I don't think these should warn by default, but I would like a suggestion.

> I'm not sure what you mean by °10? By invalid scene path, you mean something invalid after the #? That can't be known until after the asset is loaded, but we can log something in that case.

Correct, that's what I would like.

> For °11, if a system isn't running, Rust should already give a warning for something like "unused function"?

Doesn't work if your states are subtly messed up :( This is made a lot worse by the lack of a battle-tested built-in solution for "don't start the game until everything is loaded".

**Comment by mockersf** [comment]
> Disagreed here: we can check if the generated handles are to the correct type of asset upon loading. This also seems like something that could be checked based on the extension.

Not possible at all in the current asset server implementation. Maybe it will in the next.

> Doesn't work if your states are subtly messed up :( This is made a lot worse by the lack of a battle-tested built-in solution for "don't start the game until everything is loaded".

Then not an asset or rendering issue, but it could be nice to have a report like "hey, you exited the game, but those systems that have been added to a schedule never ran. maybe check their conditions and state?"

**Comment by alice-i-cecile** [comment]
> Not possible at all in the current asset server implementation. Maybe it will in the next.

Unsurprising. I think this is a pretty critical footgun and will work to fix it once the initial asset rework is merged.

> Then not an asset or rendering issue, but it could be nice to have a report like "hey, you exited the game, but those systems that have been added to a schedule never ran. maybe check their conditions and state?"

Agreed.
